File: src/MaksIT.WebUI/src/models/letsEncryptServer/account/requests/PostAccountRequest.ts
import z, { array, boolean, object, Schema, string } from 'zod'
import { ChallengeType } from '../../../../entities/ChallengeType'
import { RequestModelBase, RequestModelBaseSchema } from '../../../RequestModelBase'

export interface PostAccountRequest extends RequestModelBase {
  description: string
  contacts: string[]
  challengeType: string
  hostnames: string[]
  isStaging: boolean
  agreeToS: boolean
}

export const PostAccountRequestSchema: Schema<PostAccountRequest> = RequestModelBaseSchema.and(
  object({
    description: string(),
    contacts: array(string()),
    hostnames: array(string()),
    challengeType: z.enum(ChallengeType),
    isStaging: boolean(),
    agreeToS: boolean()
  })
)

File: src/MaksIT.Models/LetsEncryptServer/Account/Requests/PostAccountRequest.cs
using System.ComponentModel.DataAnnotations;
using MaksIT.Core.Abstractions.Webapi;

namespace MaksIT.Models.LetsEncryptServer.Account.Requests;

public class PostAccountRequest : RequestModelBase {
  public required string Description { get; set; }
  public required string[] Contacts { get; set; }
  public required string ChallengeType { get; set; }
  public required string[] Hostnames { get; set; }
  public required bool IsStaging { get; set; }
  public required bool AgreeToS { get; set; }

  public override IEnumerable<ValidationResult> Validate(ValidationContext validationContext) {
    if (string.IsNullOrWhiteSpace(Description))
      yield return new ValidationResult("Description is required", [nameof(Description)]);

    if (Contacts == null || Contacts.Length == 0)
      yield return new ValidationResult("Contacts is required", [nameof(Contacts)]);

    if (Hostnames == null || Hostnames.Length == 0)
      yield return new ValidationResult("Hostnames is required", [nameof(Hostnames)]);

    if (string.IsNullOrWhiteSpace(ChallengeType) && ChallengeType != "http-01")
      yield return new ValidationResult("ChallengeType is required", [nameof(ChallengeType)]);

    if (!AgreeToS)
      yield return  new ValidationResult("You must agree to the Terms of Service", [nameof(AgreeToS)]);
  }
}
